pkg/api/datasources.go

```diff
@@ -254,16 +257,24 @@ func UpdateDataSource(c *models.ReqContext, cmd models.UpdateDataSourceCommand)
 		if errors.Is(err, models.ErrDataSourceNotFound) {
 			return response.Error(404, "Data source not found", nil)
 		}
-		return response.Error(500, "Failed to query datasources", err)
+		return response.Error(500, "Failed to query datasource", err)
 	}
 
-	dtos := convertModelToDtos(query.Result)
+	datasourceDTO := convertModelToDtos(query.Result)
+
+	if err = bus.Publish(&events.DatasourceUpdated{
```

Comment:
@marefr -- is this reasonable?   bus is deprecated, but not sure if there is anything more appropriate

Reply:
Not sure :) It's true that it's deprecated. Data sources can be updated thru provisioning as well and this code would not execute then for example.

We normally have solved this situation by keeping track of when data source last was updated and comparing that with current last updated time, basically keeping track of this when handling a request. Similar to what the instance management in SDK does. This also works with HA. Here's some examples: 
-  https://github.com/grafana/grafana/blob/7b32c5439b35add23e34fa92b5e1764c566f7592/pkg/models/datasource_cache.go#L155-L157
- https://github.com/grafana/grafana/blob/7b32c5439b35add23e34fa92b5e1764c566f7592/pkg/tsdb/sqleng/sql_engine.go#L118-L123

Not sure if streaming requests are different? You want to cancel all current streams from the browser, re-subscribe to backend plugin streams and start streams in browser again? Have you consider handling this scenario on the plugin side? Like in https://github.com/grafana/grafana-plugin-sdk-go/blob/master/backend/stream_adapter.go

Reply:
With "runStream" -- the first request to the channel will stay open as long as there are subscribers.  We tried a plugin-side reconnect when the config changes (tracking version) BUT that requires the health check request to be set and seemed more flakey.

Provisioning :thinking: do we have runtime provisioning now?  If so we could/should put the same broadcast there

Reply:
I see. What do you mean by 

>do we have runtime provisioning now?  

https://grafana.com/docs/grafana/latest/http_api/admin/#reload-provisioning-configurations

> If so we could/should put the same broadcast there

Easy answer is no due to code duplication, but quite a lot to refactor to a place where API and provisioning uses the same service.

Since we've deprecated the bus I would recommend to replace the publish of data source updated event with a call to Live/Manager service notifying it that a data source has been updated.

Reply:
> Since we've deprecated the bus I would recommend to replace the publish of data source updated event with a call to Live/Manager service notifying it that a data source has been updated.

The reason why I used bus here is that this feels very natural to do for async events. Using bus can help consolidating asynchronous events in one place (i.e. events where a caller does not really needs a response). The problem is that current `bus.Publish` is synchronous.

I believe that bus should be used for async events and replaced with HA solution to deliver events to all instances of Grafana. But this requires changing how bus works at the moment.

Replaced bus usage with direct Live calls.

> Data sources can be updated thru provisioning as well and this code would not execute then for example.

Thanks! I checked out a provisioning code - it seems that a larger refactoring required to handle events coming from it. I've put a logic to periodically handle datasource state change from within runstream.Manager. This way provisioning will eventually be handled by periodic handler and stream will be closed or re-established (and as a bonus we handle HA case for datasource change over API). This can be handled in more optimal way by issuing less SQL queries - but again requires a bit more work so maybe current approach is fine to go with and come back to improving this further soon? 

Also added datasource delete handling (closing stream without re-submitting it).